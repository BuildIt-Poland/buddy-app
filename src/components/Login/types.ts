import { AuthPayload } from '../../../server/src/generated/schema-types';

export interface FormData {
  email: string;
  password: string;
}

export interface ErrorDialog {
  isOpen: Boolean;
  message: string;
}

export type AuthData = {
  login: AuthPayload;
};
