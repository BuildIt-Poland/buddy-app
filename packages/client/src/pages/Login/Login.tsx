import React from 'react';
import { Link } from 'react-router-dom';
import TextField from '@material-ui/core/TextField';
import Grid from '@material-ui/core/Grid';
import { makeStyles } from '@material-ui/core/styles';
import CircularProgress from '@material-ui/core/CircularProgress';
import { useForm } from 'react-hook-form';
import AuthContainer from 'atoms/AuthContainer';
import { useDialog } from 'contexts/DialogContext';
import { useAuth } from 'contexts/AuthContext';
import RoundedButton from 'atoms/RoundedButton';
import { ROUTES } from 'shared/routes';
import DICTIONARY from './dictionary';
import { FormData } from './types';

const useStyles = makeStyles(theme => ({
  form: {
    width: '100%',
  },
  submit: {
    margin: theme.spacing(3, 0),
  },
  forgotPassword: {
    fontSize: '1.5rem',
    '&:hover': {
      color: theme.palette.action.active,
    },
  },
}));

const Login: React.FC = () => {
  const classes = useStyles();
  const { register, errors, handleSubmit } = useForm<FormData>();
  const { loading, login } = useAuth();
  const { showDialog } = useDialog();

  const onFormChange = (e: React.ChangeEvent<HTMLFormElement>) =>
    (e.target.value = e.target.value.trim());

  const onSubmit = async ({ email, password }: FormData) => {
    try {
      await login(email, password);
    } catch (error) {
      if (error) {
        if (error.networkError) {
          showDialog(DICTIONARY.ERRORS.NO_NETWORK);
        } else {
          showDialog(DICTIONARY.ERRORS.NO_USER_FOUND);
        }
      }
    }
  };

  return (
    <AuthContainer data-testid='login-page' title={DICTIONARY.TITLE}>
      <form
        className={classes.form}
        data-testid='form'
        noValidate
        onChange={onFormChange}
        onSubmit={handleSubmit(onSubmit)}>
        <TextField
          inputProps={{ 'data-testid': 'email' }}
          inputRef={register({
            required: DICTIONARY.EMAIL.REQUIRED,
            pattern: {
              value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i,
              message: DICTIONARY.EMAIL.INVALID,
            },
          })}
          margin={'dense'}
          fullWidth
          label={DICTIONARY.EMAIL.LABEL}
          name='email'
          autoComplete='email'
          autoFocus
          error={!!errors.email}
          helperText={(errors.email && errors.email.message) || ' '}
        />
        <TextField
          inputProps={{ 'data-testid': 'password' }}
          inputRef={register({
            required: DICTIONARY.PASSWORD.REQUIRED,
          })}
          margin={'dense'}
          fullWidth
          name='password'
          label={DICTIONARY.PASSWORD.LABEL}
          type='password'
          autoComplete='current-password'
          error={!!errors.password}
          helperText={(errors.password && errors.password.message) || ' '}
        />
        <RoundedButton
          data-testid='submit-button'
          type='submit'
          fullWidth
          variant='contained'
          color='primary'
          disabled={loading}
          className={classes.submit}>
          {loading ? (
            <CircularProgress
              data-testid='login-progress'
              variant={'indeterminate'}
              size={32}
            />
          ) : (
            DICTIONARY.SUBMIT
          )}
        </RoundedButton>
        <Grid container justify='flex-end'>
          <Link to={ROUTES.FORGOT_PASSWORD} className={classes.forgotPassword}>
            {DICTIONARY.FORGOT_PASSWORD}
          </Link>
        </Grid>
      </form>
    </AuthContainer>
  );
};

export default Login;
