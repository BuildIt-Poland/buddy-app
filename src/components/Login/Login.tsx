import React, { useState } from 'react';
import TextField from '@material-ui/core/TextField';
import Link from '@material-ui/core/Link';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import CircularProgress from '@material-ui/core/CircularProgress';
import useForm from 'react-hook-form';
import { useMutation } from '@apollo/react-hooks';
import { useHistory } from 'react-router-dom';

import RoundedButton from '../RoundedButton';
import AlertDialog from '../AlertDialog';
import { ROUTES } from '../../shared/routes';
import { ReactComponent as SpaceMan } from '../../svg/spaceman.svg';
import LOGIN_MUTATION from '../../graphql/login.graphql';
import { auth } from '../../utils';
import DICTIONARY from './login.dictionary';

const useStyles = makeStyles(theme => ({
  title: {
    marginTop: theme.spacing(8),
  },
  spaceMan: {
    display: 'block',
    margin: '0 auto',
    marginTop: theme.spacing(3),
  },
  submit: {
    margin: theme.spacing(3, 0, 2),
  },
}));

interface FormData {
  email: string;
  password: string;
}

interface ErrorDialog {
  isOpen: Boolean;
  message: string;
}

const LoginPage = () => {
  const classes = useStyles();
  const history = useHistory();
  const [errorDialog, setErrorDialog] = useState<ErrorDialog>({
    isOpen: false,
    message: '',
  });

  const { register, errors, handleSubmit } = useForm<FormData>();
  const [loginMutation, { loading }] = useMutation(LOGIN_MUTATION, {
    onCompleted: ({ login }) => {
      auth.setToken(login.token);
      history.push(ROUTES.BUDDY_SELECT_NEWBIE);
    },
  });

  const onSubmit = async ({ email, password }: FormData) => {
    setErrorDialog({
      isOpen: false,
      message: '',
    });
    try {
      await loginMutation({
        variables: {
          email,
          password,
        },
      });
    } catch (error) {
      if (error.networkError) {
        setErrorDialog({
          isOpen: true,
          message: DICTIONARY.ERRORS.NO_NETWORK,
        });
      } else {
        setErrorDialog({
          isOpen: true,
          message: DICTIONARY.ERRORS.NO_USER_FOUND,
        });
      }
    }
  };
  return (
    <>
      <Typography
        className={classes.title}
        component='h1'
        variant='h1'
        align={'center'}>
        {DICTIONARY.TITLE}
      </Typography>
      <SpaceMan className={classes.spaceMan} />
      <form noValidate onSubmit={handleSubmit(onSubmit)}>
        <TextField
          inputProps={{ 'data-testid': 'email' }}
          inputRef={register({
            required: DICTIONARY.EMAIL.REQUIRED,
            pattern: {
              value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i,
              message: DICTIONARY.EMAIL.INVALID,
            },
          })}
          margin={'dense'}
          fullWidth
          label={DICTIONARY.EMAIL.LABEL}
          name='email'
          autoComplete='email'
          autoFocus
          error={!!errors.email}
          helperText={(errors.email && errors.email.message) || ' '}
        />
        <TextField
          inputProps={{ 'data-testid': 'password' }}
          inputRef={register({
            required: DICTIONARY.PASSWORD.REQUIRED,
          })}
          margin={'dense'}
          fullWidth
          name='password'
          label={DICTIONARY.PASSWORD.LABEL}
          type='password'
          autoComplete='current-password'
          error={!!errors.password}
          helperText={(errors.password && errors.password.message) || ' '}
        />
        <RoundedButton
          data-testid='submit-button'
          type='submit'
          fullWidth
          variant='contained'
          color='primary'
          disabled={loading}
          className={classes.submit}>
          {loading ? (
            <CircularProgress
              data-testid='login-progress'
              variant={'indeterminate'}
              size={32}
            />
          ) : (
            DICTIONARY.SIGN_IN
          )}
        </RoundedButton>
        <Grid container justify='flex-end'>
          <Link href='#'>{DICTIONARY.FORGOT_PASSWORD}</Link>
        </Grid>
        {errorDialog.isOpen && (
          <AlertDialog message={errorDialog.message}></AlertDialog>
        )}
      </form>
    </>
  );
};

export default LoginPage;
