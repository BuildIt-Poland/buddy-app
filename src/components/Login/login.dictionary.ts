interface LOGIN_DICTIONARY {
  TITLE: string;
  FORGOT_PASSWORD: string;
  SIGN_IN: string;
  EMAIL: {
    LABEL: string;
    REQUIRED: string;
    INVALID: string;
  };
  PASSWORD: {
    REQUIRED: string;
    LABEL: string;
  };
  'No such user found': string;
  SERVICE_ERROR?: string;
  NO_NETWORK: string;
}

const DICTIONARY: LOGIN_DICTIONARY = {
  TITLE: 'Buddy',
  FORGOT_PASSWORD: 'Forgot password?',
  SIGN_IN: 'Sign In',
  EMAIL: {
    LABEL: 'Email Address',
    REQUIRED: 'Email address is required',
    INVALID: 'Please enter a valid email',
  },
  PASSWORD: {
    LABEL: 'Password',
    REQUIRED: 'Password is required',
  },
  'No such user found':
    'The email and password you entered did not match our records. \n Please double-check and try again.',
  NO_NETWORK: 'Please verify your internet connection',
};

export default DICTIONARY;
