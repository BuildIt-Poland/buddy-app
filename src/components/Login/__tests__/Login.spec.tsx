import React from 'react';
import { create } from 'react-test-renderer';

import Login from '../Login';
jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('../../RoundedButton/', () => 'RoundedButton');

describe('Component - Login', () => {
  test('renders correctly', () => {
    const component = create(<Login />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
