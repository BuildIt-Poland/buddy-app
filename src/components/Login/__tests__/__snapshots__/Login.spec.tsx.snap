// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Login renders correctly 1`] = `
Array [
  <Typography
    align="center"
    component="h3"
    variant="h3"
  >
    Buddy App
  </Typography>,
  <RoundedButton
    color="primary"
    fullWidth={true}
    variant="contained"
  >
    Login
  </RoundedButton>,
]
`;
